### ISSUE TITLE
[rfc] config improvements

### ISSUE THREAD

**Issue by Nytelife26** [amperser/proselint#1375]
## Relevant issues

Supersedes #997, #1364.

# To-do items

- [ ] replace the JSON config system with TOML. `tomllib` is available as a backport from `tomli` for versions below 3.11, so we don't need to worry about backwards compatibility. 
- [ ] configuration per project would be desirable. proselint could search for overrides in project directories, and use those too.
- [x] granular selection of checks (implemented in 0.16)
  - [x] the ability to select individual checks must be implemented
  - [x] partial paths should match all checks underneath that path, i.e. setting `misc` to `true` should enable all checks within `misc`
  - [x] granular overrides should apply, i.e. setting `misc.capitalization` to `false` should deactivate all checks within `misc.capitalization`, and this should work at any level.
- [ ] replace merging system
  - [ ] use recursive explicit merging via an `extends` key ([proposal](https://github.com/amperser/proselint/pull/1468#pullrequestreview-3647262861))
  - [ ] reverse searching priority, and stop when first usable config is found
  - [ ] add rulesets for use in `extends` (#1406, #381)

**Comment by barrett-ruth** [comment]
In this, can you please allow the ignoring if specific rules - i.e. *just* `typography.symbols.curly_quotes`, and not forcing users to ignore all of  `typography.symbols`? Most linters/formatters permit this functionality.

**Comment by Nytelife26** [comment]
> In this, can you please allow the ignoring if specific rules - i.e. _just_ `typography.symbols.curly_quotes`, and not forcing users to ignore all of `typography.symbols`? Most linters/formatters permit this functionality.

the configuration structure in the refactoring effort is still being worked on. i'm still not sure if a rust port is the best idea, because it will be difficult to make it work with the current structure, but that is the goal, yes.

ultimately, efforts have been temporarily postponed while i finish a university project. i would welcome input or contributions from the community, though.

**Comment by barrett-ruth** [comment]
Sounds great. Thanks so much for your effort! (And yes, I'd definitely only rewrite it in rust if a) you have a reason to learn rust and b) it eases future development by a significant amount).

**Comment by Nytelife26** [comment]
> I'd definitely only rewrite it in rust if a) you have a reason to learn rust

oh, worry not, i already know it. i'm fairly proficient with it, and i've been writing rust for just over 3 years now.

> b) it eases future development by a significant amount).

most things that aren't python would do this. rust is a tricky one, but typescript was my other idea, and i'm not sure that's much better than python all things considered.

fortunately, i have fairly substantial experience with whole-project ports and refactors across languages. the primary motivation for doing so is that i cannot use the refactoring efforts made in python anymore, and i am determined not to let that put a stop to work on the project.

**Comment by Nytelife26** [comment]
@barrett-ruth the functionality you requested is now implemented in #1371, along with general path segment matching. for example, you can use the following to enable the entire module.
```python
    "typography": True,
```
granular overrides are something I would like to implement down the road (i.e. setting `typography` to `True` but `typography.symbols.curly_quotes` to `False` would enable everything in `typography` except for that one check), but they are not yet implemented.

**Comment by Nytelife26** [comment]
Granular selection of checks is now available via #1396
